### PR TITLE
Kill Variable

### PR DESCRIPTION
**Pull request by Balandat**
This is d-e-d. See #110

### CODE REVIEW

File: gpytorch/functions/__init__.py [hunk 1]
```diff
@@ -15,16 +16,16 @@ def add_diag(input, diag):
...
         - diag (scalar) - Scalar s so that s*I is added to the input matrix.
 
     Returns:
-        - matrix nxn - Variable or LazyTensor wrapping a new matrix with the diagonal \
+        - matrix nxn - Tensor or LazyTensor wrapping a new matrix with the diagonal \
                        component added.
     """
-    if not isinstance(diag, Variable):
-        raise RuntimeError("Expected a variable for the diagonal component.")
+    if not isinstance(diag, Tensor):
```
Comment (gpleiss) [inline gpytorch/functions/__init__.py]: Should we be using `torch.is_tensor` here instead? I remember there being some issues (back with PyTorch 0.2) with using `isinstance` with Tensors, but not sure if those have been fixed or not.

File: gpytorch/lazy/block_diagonal_lazy_tensor.py [hunk 2]
```diff
@@ -121,8 +121,8 @@ def diag(self):
     def mul(self, other):
         # We're using a custom method here - the constant mul is applied to the base_lazy tensor
         # This preserves the sum batch structure
-        if not (isinstance(other, Variable) or isinstance(other, LazyTensor)) or (
-            isinstance(other, Variable) and other.numel() == 1
+        if not (isinstance(other, Tensor) or isinstance(other, LazyTensor)) or (
```
Comment (gpleiss) [inline gpytorch/lazy/block_diagonal_lazy_tensor.py]: Same here

File: gpytorch/lazy/interpolated_lazy_tensor.py [hunk 3]
```diff
@@ -529,8 +529,8 @@ def matmul(self, tensor):
     def mul(self, other):
         # We're using a custom method here - the constant mul is applied to the base_lazy tensor
         # This preserves the interpolated structure
-        if not (isinstance(other, Variable) or isinstance(other, LazyTensor)) or (
-            isinstance(other, Variable) and other.numel() == 1
+        if not (isinstance(other, Tensor) or isinstance(other, LazyTensor)) or (
```
Comment (gpleiss) [inline gpytorch/lazy/interpolated_lazy_tensor.py]: Same here

File: gpytorch/lazy/lazy_tensor.py [hunk 4]
```diff
@@ -648,7 +643,7 @@ def mul(self, other):
             :obj:`gpytorch.lazy.ConstantMulLazyTensor`. If other was
             another matrix, this will likely be a :obj:`gpytorch.lazy.MulLazyTensor`.
         """
-        if not (isinstance(other, Variable) or isinstance(other, LazyTensor)) or (
+        if not (isinstance(other, Tensor) or isinstance(other, LazyTensor)) or (
```
Comment (gpleiss) [inline gpytorch/lazy/lazy_tensor.py]: Same here

File: gpytorch/random_variables/categorical_random_variable.py [hunk 5]
```diff
@@ -20,11 +20,11 @@ def __init__(self, mass_function):
             variables.
 
         Params:
-        - mass_function (Variable: vector k or matrix n x k) weights of categorical distribution
+        - mass_function (Tensor: vector k or matrix n x k) weights of categorical distribution
         """
         super(CategoricalRandomVariable, self).__init__(mass_function)
-        if not isinstance(mass_function, Variable):
-            raise RuntimeError("mass_function should be a Variable")
+        if not isinstance(mass_function, Tensor):
```
Comment (gpleiss) [inline gpytorch/random_variables/categorical_random_variable.py]: Same here

File: gpytorch/random_variables/gaussian_random_variable.py [hunk 6]
```diff
@@ -17,17 +17,17 @@ class GaussianRandomVariable(RandomVariable):
...
-    - covar (Variable: matrix n x n or batch matrix b x n x n) covariance of Gaussian distribution
+    - mean (Tensor: vector n or matrix b x n) mean of Gaussian distribution
+    - covar (Tensor: matrix n x n or batch matrix b x n x n) covariance of Gaussian distribution
     """
 
     def __init__(self, mean, covar):
         super(GaussianRandomVariable, self).__init__(mean, covar)
-        if not isinstance(mean, Variable) and not isinstance(mean, LazyTensor):
-            raise RuntimeError("The mean of a GaussianRandomVariable must be a Variable")
+        if not isinstance(mean, Tensor) and not isinstance(mean, LazyTensor):
```
Comment (gpleiss) [inline gpytorch/random_variables/gaussian_random_variable.py]: Same here, and right below

File: gpytorch/random_variables/samples_random_variable.py [hunk 7]
```diff
@@ -3,24 +3,24 @@
...
+        samples should be a Tensor, with the first dimension representing the samples
 
         Params:
-        - samples (Variable: b x ...) samples
+        - samples (Tensor: b x ...) samples
         """
         super(SamplesRandomVariable, self).__init__(samples)
-        if not isinstance(samples, Variable):
-            raise RuntimeError("samples should be a Variable")
+        if not isinstance(samples, Tensor):
```
Comment (gpleiss) [inline gpytorch/random_variables/samples_random_variable.py]: Same here

File: gpytorch/functions/__init__.py (hunk 1 above)
Comment (Balandat) [reply gpytorch/functions/__init__.py]: yeah let's use `torch.is_tensor`

**Pull request merged**